lib/Platform/Intl/PlatformIntlApple.mm

```diff
@@ -26,6 +26,168 @@
 }
 }
 
+// https://unicode.org/reports/tr35/#Canonical_Unicode_Locale_Identifiers
+// https://tc39.es/ecma402/#sec-canonicalizeunicodelocaleid
+vm::CallResult<std::u16string> canonicalizeLocaleId(const std::u16string &localeId, vm::Runtime *runtime) {
+  llvh::Optional<bcp47_parser::ParsedLocaleIdentifier> parserOpt = bcp47_parser::parseLocaleId(localeId);
+
+  if (!parserOpt.hasValue()) {
+    // Not structurally valid language tag
+    return runtime->raiseRangeError(vm::TwineChar16("Unicode locale id ") + vm::TwineChar16(localeId.c_str()) +vm::TwineChar16(" is not structurally valid."));
+  }
+
+  std::u16string canoLocaleId;
+  canoLocaleId.reserve(localeId.length());
+
+  // Casing, variant, and attribute sorting are handled in parsing
+  bcp47_parser::ParsedLocaleIdentifier parsedId = parserOpt.getValue();
```

Comment:
Move is more efficient here

```suggestion
  bcp47_parser::ParsedLocaleIdentifier parsedId = std::move(*parserOpt);
```

It would be even more efficient to never make a new object at all:
```suggestion
  const auto &parsedId = *parserOpt;
```